### PR TITLE
DAOS-1507 test: dmg storage scan for SCM

### PR DESCRIPTION
**Pull request by shimizukko**
This test partially covers the following requirement.
(TR-1.0.34) admin can use daos_shell to collect information and create yaml
file by himself. This means that daos_shell allows to list:
SCM module and NVMe SSDs with NUMA affinity
network adaptor with NUMA affinity

This test focuses on the correctness of SCM info obtained by dmg storage
scan (so that the admin can create a yaml file correctly). First, it
verifies the SCM Namespaces exist in /dev. Second, it verifies the namespace
count by comparing against the number of namespace rows obtained with
--verbose.

One major code update is adding regex for storage_scan and
storage_scan_verbose in dmg_utils.py.

Test-tag-hw-small: pr,hw,small dmg_storage_scan_scm
Signed-off-by: Makito Kano <makito.kano@intel.com>

### CODE REVIEW

**Comment by shimizukko** [comment]
Functional_Hardware_Medium / 01-./daos_test/daos_core_test.py:DaosCoreTest.test_subtest;test_d-num_clients-num_replicas-hosts-pool-0-1-f9f9 – DaosCoreTest -> DAOS-5325
Functional_Hardware_Medium / 12-./daos_test/daos_core_test-rebuild.py:DaosCoreTestRebuild.test_rebuild;test_r_25-num_clients-num_replicas-hosts-0-1-ca6b – DaosCoreTestRebuild -> DAOS-5315
Functional_Hardware_Medium / 13-./daos_test/daos_core_test-rebuild.py:DaosCoreTestRebuild.test_rebuild;test_r_26-num_clients-num_replicas-hosts-0-1-6d39 – DaosCoreTestRebuild -> DAOS-5315

File: src/tests/ftest/control/dmg_storage_scan_scm.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+hosts:
+  test_servers:
+    - server-A
+timeout: 102
```
Comment (amjustin13) [inline src/tests/ftest/control/dmg_storage_scan_scm.yaml:4]: Is there a reason why the timeout is so specific?

Comment (shimizukko) [reply src/tests/ftest/control/dmg_storage_scan_scm.yaml:4]: I set it to 1.5 times of the actual execution time. Maybe I should round up to the next 10s.

File: src/tests/ftest/util/dmg_utils.py [hunk 2]
```diff
@@ -74,7 +75,61 @@ def storage_scan(self, verbose=False):
...
+            while i < len(vals):
+                data["pci_addrs"].append(vals[i][5])
+                data["models"].append("{} {}".format(vals[i][6], vals[i][7]))
+                data["fw_revisions"].append(vals[i][8])
+                data["nvme_socket_ids"].append(vals[i][9])
+                data["nvme_capacity_vals"].append(vals[i][10])
+                data["nvme_capacity_units"].append(vals[i][11])
+                i += 1
+
+            return data
```
Comment (phender) [inline src/tests/ftest/util/dmg_utils.py:114]: Ideally this method would have a single return statement.

File: src/tests/ftest/util/dmg_utils.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 from grp import getgrgid
 from pwd import getpwuid
 import re
+from collections import defaultdict
```
Comment (phender) [inline src/tests/ftest/util/dmg_utils.py:30]: This brings up an interesting discussion.  If we don't have all of the expected data, won't returning an empty dictionary be sufficient or even preferable?  It would more clearly indicate something was wrong with the command output processing (`DmgCommand.result.exit_status == 0 and <return_value> is None`).

File: src/tests/ftest/util/dmg_utils.py [hunk 4]
```diff
@@ -74,7 +75,61 @@ def storage_scan(self, verbose=False):
         self.set_sub_command("storage")
         self.sub_command_class.set_sub_command("scan")
         self.sub_command_class.sub_command_class.verbose.value = verbose
-        return self._get_result()
+        self.result = self._get_result()
+        self.log.info("--- self.result.stdout ---")
+        self.log.info(self.result.stdout)
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py:79]: Yes, but it doesn't have the metadata on the left, so it makes it easier to test the regex. It's extra, but job.log for this test has near 600 lines, so the impact is small.

File: src/tests/ftest/util/dmg_utils.py [hunk 5]
```diff
@@ -74,7 +75,61 @@ def storage_scan(self, verbose=False):
...
+
+        if verbose:
+            vals = re.findall(
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py:84]: Thanks for suggestion. I implemented as you described. For the hostname, we can't directly use NodeSet; we'll have to create a class and implement some methods like __hash__ to make the hashing work with NodeSet. I used string because it would be easier to access. Also, I think the caller should consider using NodeSet to obtain the values they want.

File: src/tests/ftest/util/dmg_utils.py (hunk 2 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Done. Although not sure what's the difference.

File: src/tests/ftest/util/dmg_utils.py (hunk 3 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: I think you have a good point. The regex I wrote only works with correctly formed output. If it's not, for example no hardware storage, then it returns a partially filled dictionary. At this point, I would say I provide a utility function that covers general case only because the use cases aren't clear.

File: src/tests/ftest/util/dmg_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/dmg_utils.py:79]: We could disable the verbose mode (`DmgCommand.verbose = False`) to prevent the duplicate information, but the few additional lines may not make a big difference.

File: src/tests/ftest/util/dmg_utils.py [hunk 6]
```diff
@@ -74,7 +74,75 @@ def storage_scan(self, verbose=False):
...
+                host = row[0]
+                data[host] = {}
+                scm_total = "{} {}".format(row[1], row[2])
+                namespace = row[3]
+                nvme_total = "{} {}".format(row[4], row[5])
+                controller = row[6]
+                data[host]["scm_total"] = scm_total
+                data[host]["namespace"] = namespace
+                data[host]["nvme_total"] = nvme_total
+                data[host]["controller"] = controller
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Done.

File: src/tests/ftest/control/dmg_storage_scan_scm.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        RC_SUCCESS = 0
+        for pmem_name in pmem_names:
+            lscmd = "{} {}".format("ls", os.path.join("/dev", pmem_name))
+            # rc is a dictionary where return code is the key.
+            rc = pcmd(hosts=self.hostlist_servers, command=lscmd)
+            self.assertTrue(RC_SUCCESS in rc)
+
+        # Call without verbose and verify the namespace value.
+        data = self.get_dmg_command().storage_scan()
+        self.assertEqual(int(data[host]["namespace"]), len(pmem_names))
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/control/dmg_storage_scan_scm.py]: Done.

File: src/tests/ftest/util/dmg_utils_base.py [hunk 8]
```diff
@@ -55,6 +55,8 @@ def __init__(self, path, yaml_cfg=None):
         self.debug = FormattedParameter("-d", False)
         self.json = FormattedParameter("-j", False)
 
+        self.result = None
+
```
Comment (phender) [inline src/tests/ftest/util/dmg_utils_base.py:59]: This is no longer required as it is defined in the inherited `YamlCommand`->`SubProcessCommand`->`CommandWithSubCommand` class: https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/command_utils.py#L414

File: src/tests/ftest/util/dmg_utils.py [hunk 9]
```diff
@@ -102,20 +102,109 @@ def network_scan(self, provider=None, all_devs=False):
...
+            self.log.info("--- Non-verbose output parse result ---")
+            self.log.info(vals)
+
+            data = {}
+            for row in vals:
+                host = row[0]
+                data[host] = {
+                    "scm": {"capacity": None, "details": None},
+                    "nvme": {"capacity": None, "details": None}}
+                data[host]["scm"]["capacity"] = "{} {}".format(row[1], row[2])
```
Comment (phender) [inline src/tests/ftest/util/dmg_utils.py:202]: Optional, this could be simplified (also aligns with the first bullet in https://www.python.org/dev/peps/pep-0008/#programming-recommendations):
```suggestion
                data[host]["scm"]["capacity"] = " ".join(row[1:3])
```

File: ci/functional/test_main.sh [hunk 10]
```diff
@@ -33,7 +33,7 @@ mkdir -p install/lib/daos/TESTING/ftest/avocado/job-results
 if $TEST_RPMS; then
     # shellcheck disable=SC2029
     ssh -i ci_key -l jenkins "${first_node}" \
-      "TEST_TAG=$test_tag                            \
+      "TEST_TAG=\"$test_tag\"                        \
        TNODES=$tnodes                                \
        FTEST_ARG=$FTEST_ARG                          \
```
Comment (phender) [inline ci/functional/test_main.sh:38]: This will probably be covered in another PR, but ideally we should be doing the following:
```suggestion
      "TEST_TAG=\"$test_tag\"                        \
       TNODES=\"$tnodes\"                            \
       FTEST_ARG=\"$FTEST_ARG\"                      \
```

File: src/tests/ftest/util/dmg_utils_base.py (hunk 8 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils_base.py]: Done. Thanks for letting me know.

File: src/tests/ftest/util/dmg_utils.py (hunk 9 above)
Comment (shimizukko) [reply src/tests/ftest/util/dmg_utils.py]: Done. Looks like this page has many useful resources.

File: ci/functional/test_main.sh (hunk 10 above)
Comment (shimizukko) [reply ci/functional/test_main.sh]: Done.

**Comment by shimizukko** [comment]
Known Test Failures:
Functional_Hardware_Small / 1-./io/hdf5.py:Hdf5.test_hdf5;transport_config-client_processes-container-transport_config-hosts-pool-servers-transport_config-test_repo-f653 – Hdf5 -> DAOS-5321
Functional_Hardware_Large / 1-./osa/osa_online_reintegration.py:OSAOnlineReintegration.test_osa_online_reintegration;properties-hosts-clientslots-iorflags-pool-servers-d898 – OSAOnlineReintegration -> DAOS-5322
